src/libmongoc/src/mongoc/mongoc-cluster.c

```diff
@@ -3532,213 +3496,292 @@ network_error_reply (bson_t *reply, mongoc_cmd_t *cmd)
+                                mcd_rpc_message *rpc,
+                                bson_t *reply,
+                                bson_error_t *error)
+{
+   BSON_ASSERT_PARAM (cluster);
+   BSON_ASSERT_PARAM (cmd);
+   BSON_ASSERT_PARAM (rpc);
+   BSON_ASSERT_PARAM (reply);
+   BSON_ASSERT_PARAM (error);
 
-      section[1].payload_type = 1;
-      section[1].payload.sequence.size = cmd->payload_size +
-                                         (int32_t) identifier_len + 1 +
-                                         (int32_t) sizeof (int32_t);
-      section[1].payload.sequence.identifier = cmd->payload_identifier;
-      section[1].payload.sequence.bson_documents = cmd->payload;
-      rpc.msg.sections[1] = section[1];
-      rpc.msg.n_sections++;
-   }
+   bool ret = false;
 
-   _mongoc_rpc_gather (&rpc, &cluster->iov);
-   _mongoc_rpc_swab_to_le (&rpc);
+   mongoc_server_stream_t *const server_stream = cmd->server_stream;
 
-   if (mongoc_cmd_is_compressible (cmd)) {
-      int32_t compressor_id =
-         mongoc_server_description_compressor_id (server_stream->sd);
+   mongoc_buffer_t buffer;
+   _mongoc_buffer_init (&buffer, NULL, 0, NULL, NULL);
 
-      TRACE (
-         "Function '%s' is compressible: %d", cmd->command_name, compressor_id);
-      if (compressor_id != -1) {
-         output = _mongoc_rpc_compress (cluster, compressor_id, &rpc, error);
-         if (output == NULL) {
-            _mongoc_bson_init_if_set (reply);
-            _mongoc_buffer_destroy (&buffer);
-            return false;
-         }
-      }
+   if (!_mongoc_buffer_append_from_stream (&buffer,
+                                           server_stream->stream,
+                                           sizeof (int32_t),
+                                           cluster->sockettimeoutms,
+                                           error)) {
+      MONGOC_DEBUG (
+         "could not read message length, stream probably closed or timed out");
+      RUN_CMD_ERR_DECORATE;
+      _handle_network_error (cluster, server_stream, error);
+      server_stream->stream = NULL;
+      network_error_reply (reply, cmd);
+      goto done;
    }
 
-   _mongoc_rpc_op_egress_inc (&rpc);
-   ok = _mongoc_stream_writev_full (server_stream->stream,
-                                    (mongoc_iovec_t *) cluster->iov.data,
-                                    cluster->iov.len,
-                                    cluster->sockettimeoutms,
-                                    error);
+   const int32_t message_length = _int32_from_le (buffer.data);
+
+   if (message_length < message_header_length ||
+       message_length > server_stream->sd->max_msg_size) {
+      RUN_CMD_ERR (MONGOC_ERROR_PROTOCOL,
+                   MONGOC_ERROR_PROTOCOL_INVALID_REPLY,
+                   "message length %" PRId32
+                   " is not within valid range of 16-%" PRId32 " bytes",
+                   message_length,
+                   server_stream->sd->max_msg_size);
```

Comment:
```suggestion
      RUN_CMD_ERR (MONGOC_ERROR_PROTOCOL,
                   MONGOC_ERROR_PROTOCOL_INVALID_REPLY,
                   "message length %" PRId32
                   " is not within valid range of %" PRId32 "-%" PRId32
                   " bytes",
                   message_header_length,
                   message_length,
                   server_stream->sd->max_msg_size);
```

Use `message_header_length` constant.